README.md
# license_plate_number

[![pub package](https://img.shields.io/pub/v/license_plate_number)](https://img.shields.io/pub/v/license_plate_number)

#### Language: [English](README.md) | [中文简体](README_CN.md) 

Flutter package for inputting chinese license plate number.

- Build-in one light style and one dark style, and the style can be changed.
- Supporting landscape.
- Plate numbers can be edited separately.

![](screenshots/screenshot_portrait.jpeg)

![](screenshots/screenshot_landscape.jpeg)



## How to use

### Install

```yaml
dependencies:
  license_plate_number: ^2.0.1
```

### Basic usage

Package includes two main widgets，`PlateInputField` and `PlateKeyboard`.

`PlateInputField` is for inputting the plate number.

```dart
Container(
  child: PlateInputField(
    placeHolder: '沪A12345',
    style: PlateStyle.light,
    inputFieldWidth: 40,
    inputFieldHeight: 54,
    onChange: (List<String> array, String value) {
      // e.g.
      // array ['沪', 'A', '1', '2', '3', '4', '5', '6']
      // value 沪A123456
    },
  );
  alignment: Alignment.center,
)
```

`PlateKeyboard` is usually not used directly. It is controlled by `PlateInputField`. But we also provide `KeyboardController` to show/hide the keyboard by yourself.

```dart
/// KeyboardController
KeyboardController _keyboardController = KeyboardController();

/// Show keyboard
_keyboardController.showKeyboard(context);

/// Hide keyboard
_keyboardController.hideKeyboard();

/// Is keyboard showing
_keyboardController.isKeyboardShowing();

/// Modify plate number
_keyboardController.plateNumber = 'XXXXXXX';
```

```dart
PlateInputField(
  /// Pass KeyboardController to PlateInputField
  keyboardController: _keyboardController,
);
```


### Styles

Two default styles was build in，but self-defined style is also supported.

```dart
/// 1. Use constructor.
PlateStyles(
  /// Text style of plate input field.
  plateInputFieldTextStyle: TextStyle(
    color: Color(0xFFFFFFFF),
    fontSize: 20,
    fontWeight: FontWeight.normal,
  ),
  /// Text style of new energy place holder.
  newEnergyPlaceHolderTextStyle: TextStyle(
    color: Color(0xFFDDDDDD),
    fontSize: 10,
    fontWeight: FontWeight.normal,
  ),
  /// Background color of input field.
  plateInputFieldColor: Color(0xFF4A4A4A),
  /// Border of input field.
  plateInputBorder: Border.fromBorderSide(BorderSide(
    color: Color(0xFFFFFFFF),
    width: 2,
    style: BorderStyle.solid,
  )),
  /// Border of focused input field.
  plateInputFocusedBorder: Border.fromBorderSide(BorderSide(
    color: Color(0xFF2196F3),
    width: 2,
    style: BorderStyle.solid,
  )),
  /// Border radius of input field.
  plateInputBorderRadius: const Radius.circular(8.0),
  /// Color of Separator between city code and numbers.
  plateSeparatorColor: Color(0xFFFFFFFF),
  
  /// Background color of keyboard.
  keyboardBackgroundColor: Color(0xFF131313),
  /// Text color of key text.
  keyboardButtonTextColor: Color(0xFFFFFFFF),
  /// Background color of keys.
  keyboardButtonColor: Color(0xFF4A4A4A),
  /// Background color of disabled keys.
  keyboardButtonDisabledColor: Color(0x994A4A4A),
);
```

```dart
/// 2. Copy with a default style.
PlateStyles.dark.copyWith(
  keyboardBackgroundColorDark: Color(0xFF000000),
);
```



CHANGELOG.md
## 1.0.0

- Initial release.

## 1.0.2

- Add documents.

## 1.0.3

- Supporting plate number of new energy car.

## 1.0.4

- Fix a bug.

## 1.1.0

- Add complete button.

## 2.0.0

- Migrate to null safety.

## 2.0.1

- Fix [issue #3](https://github.com/chenglei1986/license_plate_number/issues/3)

lib/src/plate_styles.dart
import 'package:flutter/widgets.dart';

/// 车牌输入框和键盘样式
class PlateStyles {
  const PlateStyles({
    this.plateInputFieldTextStyle = plateInputFieldTextStyleLight,
    this.newEnergyPlaceHolderTextStyle = newEnergyPlaceHolderTextStyleDark,
    this.plateInputFieldColor = plateInputFieldColorLight,
    this.plateInputBorder = plateInputBorderLight,
    this.plateInputFocusedBorder = plateInputFocusedBorderLight,
    this.plateInputBorderRadius = const Radius.circular(8.0),
    this.plateSeparatorColor = plateSeparatorColorLight,
    this.keyboardBackgroundColor = keyboardBackgroundColorLight,
    this.keyboardButtonTextColor = keyboardButtonTextColorLight,
    this.keyboardButtonColor = keyboardButtonColorLight,
    this.keyboardButtonDisabledColor = keyboardButtonDisabledColorLight,
  });

  /// 车牌号字体样式
  final TextStyle plateInputFieldTextStyle;

  /// 新能源输入框占位字符样式
  final TextStyle newEnergyPlaceHolderTextStyle;

  /// 输入框背景
  final Color plateInputFieldColor;

  /// 输入框边框
  final Border plateInputBorder;

  /// 输入框获取焦点时的边框
  final Border plateInputFocusedBorder;

  /// 输入框边框圆角半径
  final Radius plateInputBorderRadius;

  /// 省市与号码之间分隔符的颜色
  final Color plateSeparatorColor;

  /// 键盘背景色
  final Color keyboardBackgroundColor;

  /// 键盘按钮字体颜色
  final Color keyboardButtonTextColor;

  /// 键盘按钮背景色
  final Color keyboardButtonColor;

  /// 键盘按钮不可用时的背景色
  final Color keyboardButtonDisabledColor;

  /// 黑暗风格车牌号字体样式
  static const TextStyle plateInputFieldTextStyleDark = TextStyle(
    color: Color(0xFFFFFFFF),
    fontSize: 20,
    fontWeight: FontWeight.normal,
  );

  /// 黑暗风格车牌号字体样式
  static const TextStyle newEnergyPlaceHolderTextStyleDark = TextStyle(
    color: Color(0xFFDDDDDD),
    fontSize: 10,
    fontWeight: FontWeight.normal,
  );

  /// 黑暗风格输入框背景色
  static const Color plateInputFieldColorDark = Color(0xFF4A4A4A);

  /// 黑暗风格输入框边框
  static const Border plateInputBorderDark = Border.fromBorderSide(BorderSide(
    color: Color(0xFFFFFFFF),
    width: 2,
    style: BorderStyle.solid,
  ));

  /// 黑暗风格输入框获取焦点时的边框
  static const Border plateInputFocusedBorderDark =
      Border.fromBorderSide(BorderSide(
    color: Color(0xFF2196F3),
    width: 2,
    style: BorderStyle.solid,
  ));

  /// 黑暗风格分隔符的颜色
  static const Color plateSeparatorColorDark = Color(0xFFFFFFFF);

  /// 黑暗风格键盘背景色
  static const Color keyboardBackgroundColorDark = Color(0xFF131313);

  /// 黑暗风格键盘按钮字体颜色
  static const Color keyboardButtonTextColorDark = Color(0xFFFFFFFF);

  /// 黑暗风格键盘按钮背景色
  static const Color keyboardButtonColorDark = Color(0xFF4A4A4A);

  /// 黑暗风格键盘按钮不可用时的背景色
  static const Color keyboardButtonDisabledColorDark = Color(0x994A4A4A);

  /// 明亮风格车牌号字体样式
  static const TextStyle plateInputFieldTextStyleLight = TextStyle(
    color: Color(0xFF333333),
    fontSize: 20,
    fontWeight: FontWeight.normal,
  );

  /// 明亮风格车牌号字体样式
  static const TextStyle newEnergyPlaceHolderTextStyleLight = TextStyle(
    color: Color(0xFF666666),
    fontSize: 10,
    fontWeight: FontWeight.normal,
  );

  /// 明亮风格输入框背景色
  static const Color plateInputFieldColorLight = Color(0xFFFFFFFF);

  /// 明亮风格输入框边框
  static const Border plateInputBorderLight = Border.fromBorderSide(BorderSide(
    color: Color(0xFF333333),
    width: 2,
    style: BorderStyle.solid,
  ));

  /// 明亮风格输入框获取焦点时的边框
  static const Border plateInputFocusedBorderLight =
      Border.fromBorderSide(BorderSide(
    color: Color(0xFF2196F3),
    width: 2,
    style: BorderStyle.solid,
  ));

  /// 明亮风格分隔符的颜色
  static const Color plateSeparatorColorLight = Color(0xFF333333);

  /// 明亮风格键盘背景色
  static const Color keyboardBackgroundColorLight = Color(0xFFE1E1E1);

  /// 明亮风格键盘按钮字体颜色
  static const Color keyboardButtonTextColorLight = Color(0xFF333333);

  /// 明亮风格键盘按钮背景色
  static const Color keyboardButtonColorLight = Color(0xFFFFFFFF);

  /// 明亮风格键盘按钮不可用时的背景色
  static const Color keyboardButtonDisabledColorLight = Color(0x99FFFFFF);

  /// 黑暗风格
  static const PlateStyles dark = const PlateStyles(
    plateInputFieldTextStyle: plateInputFieldTextStyleDark,
    newEnergyPlaceHolderTextStyle: newEnergyPlaceHolderTextStyleDark,
    plateInputFieldColor: plateInputFieldColorDark,
    plateInputBorder: plateInputBorderDark,
    plateInputFocusedBorder: plateInputFocusedBorderDark,
    plateInputBorderRadius: const Radius.circular(8.0),
    plateSeparatorColor: plateSeparatorColorDark,
    keyboardBackgroundColor: keyboardBackgroundColorDark,
    keyboardButtonTextColor: keyboardButtonTextColorDark,
    keyboardButtonColor: keyboardButtonColorDark,
    keyboardButtonDisabledColor: keyboardButtonDisabledColorDark,
  );

  /// 明亮风格
  static const PlateStyles light = const PlateStyles(
    plateInputFieldTextStyle: plateInputFieldTextStyleLight,
    newEnergyPlaceHolderTextStyle: newEnergyPlaceHolderTextStyleLight,
    plateInputFieldColor: plateInputFieldColorLight,
    plateInputBorder: plateInputBorderLight,
    plateInputFocusedBorder: plateInputFocusedBorderLight,
    plateInputBorderRadius: const Radius.circular(8.0),
    plateSeparatorColor: plateSeparatorColorLight,
    keyboardBackgroundColor: keyboardBackgroundColorLight,
    keyboardButtonTextColor: keyboardButtonTextColorLight,
    keyboardButtonColor: keyboardButtonColorLight,
    keyboardButtonDisabledColor: keyboardButtonDisabledColorLight,
  );

  PlateStyles copyWith({
    TextStyle? plateInputFieldTextStyle,
    TextStyle? newEnergyPlaceHolderTextStyle,
    Color? plateInputFieldColor,
    Border? plateInputBorder,
    Border? plateInputFocusedBorder,
    Radius? plateInputBorderRadius,
    Color? plateSeparatorColor,
    Color? keyboardBackgroundColor,
    Color? keyboardButtonTextColor,
    Color? keyboardButtonColor,
    Color? keyboardButtonDisabledColor,
  }) {
    return PlateStyles(
      plateInputFieldTextStyle:
          plateInputFieldTextStyle ?? this.plateInputFieldTextStyle,
      newEnergyPlaceHolderTextStyle:
          newEnergyPlaceHolderTextStyle ?? this.newEnergyPlaceHolderTextStyle,
      plateInputFieldColor: plateInputFieldColor ?? this.plateInputFieldColor,
      plateInputBorder: plateInputBorder ?? this.plateInputBorder,
      plateInputFocusedBorder:
          plateInputFocusedBorder ?? this.plateInputFocusedBorder,
      plateInputBorderRadius:
          plateInputBorderRadius ?? this.plateInputBorderRadius,
      plateSeparatorColor: plateSeparatorColor ?? this.plateSeparatorColor,
      keyboardBackgroundColor:
          keyboardBackgroundColor ?? this.keyboardBackgroundColor,
      keyboardButtonTextColor:
          keyboardButtonTextColor ?? this.keyboardButtonTextColor,
      keyboardButtonColor: keyboardButtonColor ?? this.keyboardButtonColor,
      keyboardButtonDisabledColor:
          keyboardButtonDisabledColor ?? this.keyboardButtonDisabledColor,
    );
  }
}
